defines language more strictly in diagram section

ablwr opened the pull request:
Seeking to make the language used in this Diagram more strict, more concise, and more clear in scope.

retokromer commented on diagram.md:
```diff
@@ -45,7 +45,7 @@ The following diagram represents a simple Matroska file, comprised of an `EBML D
 The Matroska `EBML Schema` defines eight `Top Level Elements`: `SeekHead`, `Info`, `Tracks`, `Chapters`, `Cluster`, `Cues`, `Attachments`, and `Tags`.
 
-The `SeekHead Element` (also known as `MetaSeek`) contains an index of where other `Top Level Elements` of the `Segment` are located in order to let the parser know where the other major parts of the file are. This element isn't technicaly REQUIRED, but without a `SeekHead Element` a `Matroska Parser` would have to search the entire file to find all of the other `Top Level Elements`. This is beca ...
+The `SeekHead Element` (also known as `MetaSeek`) contains an index where other `Top Level Elements` of the `Segment`. This index defines where the other major parts of the file are. The `SeekHead Element` is RECOMMENDED. Without a `SeekHead Element`, a Matroska parser would have to search the entire file to find all of the other `Top Level Elements`. This is because Matroska has flexible orderin ...
```
I personally would put `The SeekHead Element is RECOMMENDED.` before the `This index defines where` sentence.

mbunkus replied:
There's a verb missing:

> …contains an index where other `Top Level Elements` of the `Segment`.

mbunkus replied:
> This is because Matroska has flexible ordering requirements

Sounds a bit French/German to me (I know that that's how it was written before; but why not change it now, too?); maybe something like "This is due to Matroska's flexible…"?

retokromer commented on diagram.md:
```diff
@@ -91,7 +91,7 @@ The `Info Element` contains vital information for identifying the whole `Segment
 Figure: Representation of a `Info Element` and its `Child Elements`.
 
-The `Tracks Elements` tells us the technical details of what is in each track. For instance, is it a video, audio or subtitle track? What resolution is the video? What sample rate is the audio? The `Tracks Elements` can store the name, number, unique identifier, language, and type (audio, video, subtitles, etc) of each track. The `Tracks Element` also identifies what codec to use to decode the tr ...
+The `Tracks Elements` defines the technical details for each track. The `Tracks Elements` can store the name, number, unique identifier, language, and type (audio, video, subtitles, etc) of each track. The `Tracks Element` can store information about the resolution of a video track or sample rate of an audio track. The `Tracks Element` also identifies what codec to use to decode the track and con ...
```
`subtitles, etc.)` with a period

mbunkus replied:
> The `Tracks Elements` defines…

`Elements` is plural; hence "define" and not "defines", or even better, make `Tracks Elements` singular (= `Tracks Element`, there's usually only one instance in a Document). If you make it singular, then all uses in this paragraph should be changed along with the corresponding verbs.

mbunkus replied:
> The `Tracks Element` can store…

You're starting two sentences with this. Better change one of them somehow. Even though we're not writing prose, it stands out and distracts from the content.

mbunkus replied:
> …identifier, language, and type…

In German you wouldn't use a `,` before the conjunction in an enumeration, and I think you don't in English, either. Or am I wrong?

ablwr replied:
Thanks both! You're right about the conjunction although it's a battle I still subconsciously fight. ;) Correcting these and others, thank you.

mbunkus commented on diagram.md:
```diff
@@ -191,7 +191,7 @@ The `Cluster Elements` contain all of the video frames and audio for each track.
 Figure: Representation of a `Cluster Element` and its immediate `Child Elements`.
 
-Below is a representation of the Block structure.
+Below is a representation of the `Block` structure.
```
Hmm maybe "…the `Block Element's` structure"? I don't feel too strongly about this, though.

mbunkus commented on diagram.md:
```diff
@@ -170,7 +170,7 @@ The `Chapters Element` section lists all of the Chapters. Chapters are a way to
 Figure: Representation of the `Chapters Element` and a selection of its `Descendant Elements`.
 
-The `Cluster Elements` contain all of the video frames and audio for each track. In a given Matroska file, there are usually many `Cluster Elements`. The Clusters help to break up the `SimpleBlock` or `BlockGroup Elements` and help with seeking and error protection. It is RECOMMENDED the size of each individual `Cluster Element` be limited to store no more than 5 seconds or 5 megabytes. Every Clu ...
+The `Cluster Elements` contain all of the video frames and audio for each track. In a given Matroska file, there are usually many `Cluster Elements`. A Matroska file SHOULD contain at least one `Cluster Elements`. The `Clusters` help to break up the `SimpleBlock` or `BlockGroup Elements` and help with seeking and error protection. It is RECOMMENDED that the size of each individual `Cluster` be li ...
```
> …video frames and audio for…

Audio what? I'd use "video and audio frames".

mbunkus replied:
> Every `Cluster` MUST contains a…

`Cluster` is singular, therefore: "contain"

mbunkus replied:
> There SHOULD be one or more `BlockGroups` or `SimpleBlocks`…

We usually talk about `Elements` and don't pluralize the element names; meaning: "one or more `BlockGroup` or `SimpleBlock Elements`…"

retokromer replied:
@mbunkus It’s called [Oxford comma](https://en.wikipedia.org/wiki/Serial_comma) – which I don’t like, BTW).

ablwr replied:
Audio comes as a stream and not chunked as frames, like video does. I'm more inclined to take the `frames` part out of video. Although `stream` is meaningful, I'll try to see if I can make it work here.

ablwr replied:
Thanks! It's so obvious I was going back-and-forth with trying to understand the proper way to describe singular elements defined with a plural (Tracks Element) and if they could be described in the above way. Updating now.

mbunkus replied:
@retokromer Aaaaah _that's_ what an Oxford comma is. I knew the term but somehow never bothered to find out what it actually means.

mbunkus replied:
> Audio comes as a stream and not chunked as frames, like video does.

That's completely wrong. Each and every compressed audio format uses chunks of indivisible data that are decoded to a certain number of samples. These chunks are usually called "audio frames", see e.g. [Wikipedia's article about MP3](https://en.wikipedia.org/wiki/MP3#File_structure). Only uncompressed audio might be stored sample by sample.

ablwr replied:
😬 Whoops!! 😆 Frames it is.

ablwr replied:
A new benefit to specification review -- free education!

mbunkus replied:
For me, too, see that pesky Oxford comma 😁

ablwr replied:
Hmm, and I thought the Oxford comma had recently fallen out of recommended use but this Wikipedia article indicates that it is still favorable use, here in the States for American English? But internationally, not so much? I'm gonna... take it out.

mbunkus commented on diagram.md:
```diff
@@ -208,11 +208,11 @@ Below is a representation of the Block structure.
   * Frame
 
-Although the Timecode value is stored once per Cluster, another timecode is stored within the Block structure itself. The way this works is that the Timecode in the Cluster is relative to the entire `Segment`. It is usually the Timecode that the first Block in the Cluster needs to be played at. The Timecode in the Block itself is relative to the Timecode in the Cluster. For example, let's say tha ...
+The `Timecode Element` value MUST be stored once per `Cluster`. Another `Timecode` SHOULD be stored within the `Block` structure itself. The `Timecode` in the `Cluster` is relative to the entire `Segment`. The `Timecode` SHOULD be the first `Block` in the `Cluster`. The `Timecode` in the `Block` itself is relative to the `Timecode` in the `Cluster`. For example, if the `Timecode` in the `Cluster` ...
```
> The `Timecode Element` value MUST be stored once per `Cluster`. Another `Timecode` SHOULD be stored within the `Block` structure itself.

I consider the second use of `Timecode` to be wrong. The first one is talking about the `Timecode Element`, an "unsigned integer" EBML `Element`. The second use is talking about specific parts (bits) of the `Block` element. Yes, it is a timecode, but not a `Timecode` (not an element of its own). Additionally it's actually only an offset to the `Timecode Element` stored in the cluster; adding that offset to the `Timecode Element's` value yields the `Block Element's` actual timecode.

I'd convert this section into something like I've written above, e.g. start with "Each `Cluster` MUST contain a `Timecode` element. Additionally the `Block Element` contains an offset that, when added to the `Cluster's Timecode Element's` value, yields the `Block's` effective timecode."

That offset in the `Block` element isn't optional either; therefore "SHOULD" is definitely the wrong word. Any other qualifier such as MUST or MAY is just as wrong. The `Block` element always consists of several bits that make up said offset. They're a fixture.

> The `Timecode` in the `Block` itself is relative to the `Timecode` in the `Cluster`.

The first use is wrong again as you're talking about those specific bits in the `Block Element` again, and not about a real `EBML Element`.

> …the `Timecode` in the `Block` would be set to 2 seconds.

Dito.

mbunkus replied:
> The `Timecode` SHOULD be the first `Block` in the `Cluster`.

This is wrong; you probably meant "…the first `Element`…".

ablwr replied:
Thanks! I had a hard time understanding what was meant by this section, too. This helps clear it up but please re-review when I re-submit.

mbunkus replied:
Wilco. The original wording wasn't really clear either, to be honest.

ablwr replied:
> The `Timecode Element` SHOULD be the first `Element` in the `Cluster`.

In order_guidelines.md, it says the Timecode element MUST be the first Element in the Cluster.

mbunkus commented on diagram.md:
```diff
@@ -208,11 +208,11 @@ Below is a representation of the Block structure.
+The `ReferenceBlock` in the `BlockGroup` is used instead of the basic "P-frame"/"B-frame" description. Instead of simply saying that this `Block` depends on the `Block` directly before, or directly afterwards, the `Timecode` of the necessary `Block` is used. Because there can be as many `ReferenceBlock Elements` as necessary for a `Block`, it allows for some extremely complex referencing.
 
-The `Cues Element` is used to seek when playing back a file by providing a temporal index for each of the tracks. It is similar to the `SeekHead Element`, but this is used for seeking to a specific time when playing back the file. Without this it is possible to seek, but it is much more difficult because the player has to 'hunt and peck' through the file looking for the correct timecode. `Cues` c ...
+The `Cues Element` is used to seek when playing back a file by providing a temporal index for each of the `Tracks`. It is similar to the `SeekHead Element`, but used for seeking to a specific time when playing back the file. It is possible to seek without this element, but it is much more difficult because the player has to 'hunt and peck' through the file looking for the correct timecode. `Cues` ...
```
> `Cues` MAY contain `CuePoint Elements`…

I'd prefer SHOULD here as I cannot think of any use case where empty `Cues` would be used. MUST would be too strong as I wouldn't want to invalidate a file just because there's an empty `Cues` master present.

mbunkus replied:
>…which store the timecode (`CueTime`) and a listing for the exact position in the file …

What's actually stored is the position of the `Cluster Element` that contains the `BlockGroup` or `SimpleBlock Element` the `CueTime` is referring to. Your sentence sounds a bit as if the position of the `BlockGroup` or `SimpleBlock Elements` itself is stored, which it isn't (and it wouldn't make sense to store that position either as a player has to look for the cluster's `Timecode Element` anyway).

mbunkus commented on diagram.md:
```diff
@@ -252,7 +252,7 @@ The `Attachments Element` is for attaching files to a Matroska file such as pict
 Figure: Representation of a `Attachments Element`.
 
-The `Tags Element` contains metadata that describes the `Segment` and potentially its `Tracks`, `Chapters`, and `Attachments`. Each Track or Chapter that those tags applies to has its UID listed in the tags. The Tags contain all extra information about the file, script writer, singer, actors, directors, titles, edition, price, dates, genre, comments, etc. And it allows you to enter many of these  ...
+The `Tags Element` contains metadata that describes the `Segment` and potentially its `Tracks`, `Chapters`, and `Attachments`. Each `Track` or `Chapter` that those tags applies to has its UID listed in the `Tags`. The `Tags` contain all extra information about the file: scriptwriter, singer, actors, directors, titles, edition, price, dates, genre, comments, etc. It allows you to enter many of the ...
```
> It allows you to…

Maybe change the personal form of address to something impersonal such as "the author". Sounds more RFCy.

ablwr commented:
Updated. When the time comes, I will squash my commits, but there may be more work to do before I do that.

ablwr commented on diagram.md:
```diff
@@ -91,7 +91,7 @@ The `Info Element` contains vital information for identifying the whole `Segment
 Figure: Representation of a `Info Element` and its `Child Elements`.
 
-The `Tracks Elements` tells us the technical details of what is in each track. For instance, is it a video, audio or subtitle track? What resolution is the video? What sample rate is the audio? The `Tracks Elements` can store the name, number, unique identifier, language, and type (audio, video, subtitles, etc) of each track. The `Tracks Element` also identifies what codec to use to decode the tr ...
+The `Tracks Element` defines the technical details for each track and can store the name, number, unique identifier, language and type (audio, video, subtitles, etc.) of each track. The `Tracks Element` can store information about the resolution of a video track or sample rate of an audio track. The `Tracks Element` also identifies what codec to use to decode the track and contains the codec's pr ...
```
"The `Tracks Element` also identifies what codec to use to decode the track and contains the codec's private data for the track."

@mbunkus Should this be worded as a can-identify, or is this a MUST or a SHOULD? I'm going to change the wording of this to the former unless it's one of the latter.

mbunkus replied:
The `Tracks Element` MUST identify all the data needed by the codec to decode the data properly. Which pieces of information are needed depends on the codec in question, e.g. the audio bit rate is only needed for uncompressed audio but not for e.g. AC-3; the track language is completely optional for all cases and the `Codec Element` MUST be present for all tracks as its the primary identifier for which decoder to use.

How you turn that into proper standards lingo is an exercise I'll gladly leave to you 😄

ablwr replied:
I did my best!

ablwr commented:
Why is the `Block Element` figure in a list form but most of the others are in a table format? Should the Block figure be converted to table, or should the tables become lists?

mbunkus commented:
> Why is the Block Element figure in a list form but most of the others are in a table format? Should the Block figure be converted to table, or should the tables become lists?

Personally I prefer the tables over the lists, especially when trying to convey multiple pieces of information per element (e.g. position, size in bits, meaning…).

ablwr commented:
@mbunkus Squashed!

retokromer commented on diagram.md:
```diff
@@ -89,9 +89,11 @@ The `Info Element` contains vital information for identifying the whole `Segment
-The `Tracks Elements` tells us the technical details of what is in each track. For instance, is it a video, audio or subtitle track? What resolution is the video? What sample rate is the audio? The `Tracks Elements` can store the name, number, unique identifier, language, and type (audio, video, subtitles, etc) of each track. The `Tracks Element` also identifies what codec to use to decode the tr ...
+The `Tracks Element` defines the technical details for each track and can store the name, number, unique identifier, language and type (audio, video, subtitles, etc.) of each track. For example, the `Tracks Element` MAY store information about the resolution of a video track or sample rate of an audio track.
+
+The `Tracks Element` MUST identify all the data needed by the codec to decode the data of the specified track. However, the data required is contingent on the codec used for the track. For example, a `Track Element` for uncompressed audio only requires the audio bit rate to be present. A codec such as AC-3 would require that the `CodecID Element` be present for all tracks, as it is the primary wa ...
```
Why is `Tracks Element` split into two paragraphs?

ablwr replied:
I probably did this for readability but nothing beyond that.

retokromer replied:
My question was about usability (i.e. find all information at one place)… but nothing beyond that!

robUx4 commented on diagram.md:
```diff
@@ -191,28 +193,39 @@ The `Cluster Elements` contain all of the video frames and audio for each track.
+|       | Header     | TrackNumber |
+|       |            |-------------|
+|       |            | Timecode    |
+|       |            |------------------------+
```
Not too sure about the extra column. It's confusing.
You could just put Flags and below: Gap (1 bit), lacing (1 bit).

ablwr replied:
Or not change to a column at all, revert back to list?

mbunkus replied:
I like diagrams, but I agree that the "bit flag" column makes this a bit ambiguous. We could combine both approaches:

* Leave the diagram in but remove the "bit flag" column
* Describe the flags in a sentence or two below the diagram

robUx4 commented on diagram.md:
```diff
@@ -191,28 +193,39 @@ The `Cluster Elements` contain all of the video frames and audio for each track.
+The `ReferenceBlock` in the `BlockGroup` is used instead of the basic "P-frame"/"B-frame" description. Instead of simply saying that this `Block` depends on the `Block` directly before, or directly afterwards, the `Timecode` of the necessary `Block` is used. Because there can be as many `ReferenceBlock Elements` as necessary for a `Block`, it allows for some extremely complex referencing.
 
-Although the Timecode value is stored once per Cluster, another timecode is stored within the Block structure itself. The way this works is that the Timecode in the Cluster is relative to the entire `Segment`. It is usually the Timecode that the first Block in the Cluster needs to be played at. The Timecode in the Block itself is relative to the Timecode in the Cluster. For example, let's say tha ...
+The `Cues Element` is used to seek when playing back a file by providing a temporal index for each of the `Tracks`. It is similar to the `SeekHead Element`, but used for seeking to a specific time when playing back the file. It is possible to seek without this element, but it is much more difficult because the player has to 'hunt and peck' through the file looking for the correct timecode.
```
It's not "each of the Tracks" but only "some of the Tracks", usually on the video track(s).

robUx4 replied:
Also that might be a good place to define if we have to add all Cuepoints for a track or only a few for that track. IMO we should recommend that video track(s) index at least all their keyframes. For audio maybe the first Block in each Cluster.

ablwr replied:
That point comes up two paragraphs down, should it be moved up?

mbunkus commented on diagram.md:
```diff
@@ -252,7 +265,7 @@ The `Attachments Element` is for attaching files to a Matroska file such as pict
 Figure: Representation of a `Attachments Element`.
 
-The `Tags Element` contains metadata that describes the `Segment` and potentially its `Tracks`, `Chapters`, and `Attachments`. Each Track or Chapter that those tags applies to has its UID listed in the tags. The Tags contain all extra information about the file, script writer, singer, actors, directors, titles, edition, price, dates, genre, comments, etc. And it allows you to enter many of these  ...
+The `Tags Element` contains metadata that describes the `Segment` and potentially its `Tracks`, `Chapters`, and `Attachments`. Each `Track` or `Chapter` that those tags applies to has its UID listed in the `Tags`. The `Tags` contain all extra information about the file: scriptwriter, singer, actors, directors, titles, edition, price, dates, genre, comments, etc. Multiple tags (title, edition, com ...
```
> Multiple tags (title, edition, comments, etc.) can be entered in different languages.

I'm not happy with the wording "Multiple tags" here. Tags are key/value pairs. The intent is to be able to store multiple instances of the value portion for each key. Maybe something like this?

> Tags can contain their values in multiple languages. For example, a movie's "title" `Tag` might contain both the English original as well as the name it was released as in Germany.

mbunkus commented on diagram.md:
```diff
@@ -191,28 +193,39 @@ The `Cluster Elements` contain all of the video frames and audio for each track.
+The `Cues Element` SHOULD contain at least one `CuePoint Element`. Each `CuePoint Element` stores the position of the `Cluster` that contains the `BlockGroup` or `SimpleBlock Element`. The timecode is stored in the `CueTime Element` and location is stored in the `CueTrackPositions Element`.
 
-The `Cues Element` is used to seek when playing back a file by providing a temporal index for each of the tracks. It is similar to the `SeekHead Element`, but this is used for seeking to a specific time when playing back the file. Without this it is possible to seek, but it is much more difficult because the player has to 'hunt and peck' through the file looking for the correct timecode. `Cues` c ...
+The `Cues Element` is flexible. For instance, `Cues Element` can be used to index every single timecode of every `Block` or can be indexed selectively. For video files, it is RECOMMENDED to index at least the keyframes of the video track.
```
> …or can be indexed selectively.

Isn't that missing a "they" before "can"?

ablwr replied:
I will add it! I don't think of it as being required but it does help with clarity.

ablwr commented on diagram.md:
````diff
@@ -191,28 +193,39 @@ The `Cluster Elements` contain all of the video frames and audio for each track.
+```
+Figure: Representation of the `Block Element` structure.  
+
+Each `Cluster` MUST contain a `Timecode Element`. The `Timecode Element` value MUST be stored once per `Cluster`. The `Timecode Element` in the `Cluster` is relative to the entire `Segment`. The `Timecode Element` SHOULD be the first `Element` in the `Cluster`.
````
+1

ablwr commented:
Updated again, thanks @mbunkus. Opinions on the diagram around code line 201 and @robUx4's comments?

mbunkus commented:
For some reason you've accidentally pushed a second commit "Add Tags as supported elements in webm." which is [already part of the repo anyway](3e20e36818535b841edeeb9723b16567e3d5e6ae).

ablwr commented:
Ah! I merged in `master` by accident to rebase and pull in any conflicting changes. 🤔

ablwr commented:
I dropped it right back out! `git rebase -i` was acting up for me yesterday (which is to say, my brain was acting up). I simplified the diagram, let me know what you think of it.

mbunkus commented on diagram.md:
```diff
@@ -141,7 +143,7 @@ The `Tracks Elements` tells us the technical details of what is in each track. F
 Figure: Representation of the `Tracks Element` and a selection of its `Descendant Elements`.
 
-The `Chapters Element` section lists all of the Chapters. Chapters are a way to set predefined points to jump to in video or audio.
+The `Chapters Element` lists all of the Chapters. Chapters are a way to set predefined points to jump to in video or audio.
```
I'd downcase the "Chapters" at the end of the sentence as you're not talking about the element itself but about the concept.

ablwr replied:
OK we're almost there!!!!!!

mbunkus commented on diagram.md:
````diff
@@ -191,28 +193,38 @@ The `Cluster Elements` contain all of the video frames and audio for each track.
+|       |--------------------------|
+|       | Data       | Frame       |  
++----------------------------------+
+```
````
I like!

ablwr commented:
OK I'm in a squash-battle again, let me smush it down one.. last.. time.

ablwr commented:
*Finally!* I rebased poorly and had to resolve destroying a merge-commit flavored commit. 😵

retokromer commented:
Are you feeling well? ;-)

ablwr commented:
Hey @retokromer! Hahaha. OK, I will give @robUx4 a chance to look it over and then MERGE IT IN!!!!!!!!

ablwr commented:
Good to merge? @mbunkus will you? Anything else that comes up can be dealt with in subsequent PRs.

mbunkus commented:
The eternal question: will he, won't he…
